src/rpc/blockchain.cpp

```diff
@@ -1710,37 +1710,61 @@ static UniValue getblockstats(const UniValue &params, bool fHelp)
 
     LOCK(cs_main);
 
-    CBlockIndex *pindex;
-    if (params[0].isNum())
-    {
-        const int height = params[0].get_int();
-        const int current_tip = chainActive.Height();
-        if (height < 0)
+    CBlockIndex *pindex = nullptr;
+    bool isNumber = true;
+    int height = -1;
+    if (!params[0].isNum())
+    {
+        // determine if string is the height or block hash
```

Comment:
Although this algorithm works in practice, I don't love it because in theory a block could be created who's hash is all 0-9 digits.  We can easily find an approximate probability of this by figuring the probability of a hash which is all digits 0-7 since that corresponds to groups of 4 bits that are 0xxx.  So 1/4 of the bits (or 64 out of 256 need to be 0), resulting in a probability of 1 in 2^^64).  This is what I mean by "works in practice".

But code gets cut and paste sometimes without careful analysis, so the code could be pasted into a context where this probability does not hold (if, for example the passed hex was not a cryptographically secure hash).  At a minimum this needs a comment describing what's going on, doing the above analysis and warning people.

But we also do this in "getblock" so I checked how it works: it looks up the "hash" in the block index and if there's a miss, it assumes its a number.  This isn't great because its a linear lookup.

But this analysis made me realize that there's a simple determinant: string length.
>>> len(str(2**64-1))
20

So the biggest 64 bit number is only 20 characters but hashes are 64 characters.

Hashes are not strictly numbers in the sense that leading 0s cannot be dropped, so I think that this is the best determinant.  It eliminates a while loop, and would allow us to accept hex integers sometime in the future.